singlecellmultiomics/bamProcessing/split_bam_by_cluster.py

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python
+'''
+DESCRIPTION
+
+    Split bams into pseudobulks based on metadata
+
+FOR HELP
+
+    python split_bam_by_cluster.py --help
+
+AUTHOR:      Jake Yeung (j.yeung@hubrecht.eu)
+LAB:         Quantitative Biology Lab (https://www.hubrecht.eu/research-groups/van-oudenaarden-group/)
+CREATED ON:  2020-01-09
+LAST CHANGE: see git log
+LICENSE:     MIT License (see: http://opensource.org/licenses/MIT)
+'''
+
+import sys, argparse, datetime
+import pysam
+import csv
+import os
+
+def reformat_header(header, add_prefix="chr"):
+    new_header = header.to_dict()
+    contigs = new_header['SQ']
+    new_contigs = []
+    for contig in contigs:
+        contig['SN'] = ''.join([add_prefix, contig['SN']])
+        new_contigs.append(contig)
+    new_header['SQ'] = new_contigs
+    return pysam.AlignmentHeader.from_dict(new_header)
+
+def scan_bam_get_sampnames(infile, tag="SM"):
```

Comment:
To prevent iterating the bam file multiple times it is possible to extract the cells in the bam file from the RG section in the header like:
```
with pysam.AlignmentFile('./tagged.bam') as f:
    print( set([entry['SM'] for entry in f.header.as_dict()['RG']]) )
```